.github/workflows/action-format.yml

```diff
@@ -0,0 +1,130 @@
+        run: |
+          mkdir ~/.ssh
+          echo "${{ secrets.DEPLOY_KEY }}" > ~/.ssh/id_ed25519
+          chmod 600 ~/.ssh/id_ed25519
+
+      - name: 'Checkout code'
+        run: |
+          PR_DATA="/tmp/pr.json"
+
+          HEAD_REF=$(jq -r ".head.ref" "$PR_DATA")
+
+          if [ ${{ steps.fork_status.outputs.fork }} == "false" ]; then
+            echo "::debug::Setting up repo using SSH"
+            HEAD_REPO=$(jq -r '.head.repo.ssh_url' "$PR_DATA")
+          else
+            echo "::debug::Setting up repo using HTTPS"
+            HEAD_REPO=$(jq -r '.head.repo.clone_url | sub("https://"; "https://x-access-token:${{ secrets.GITHUB_TOKEN }}@")' "$PR_DATA")
+          fi
+
+          git clone $HEAD_REPO .
+          git checkout -b "$HEAD_REF" "origin/$HEAD_REF"
+
+      - name: Use Node.js LTS (16.x)
+        uses: actions/setup-node@1f8c6b94b26d0feae1e387ca63ccbdc44d27b561
+        with:
+          node-version: '16'
+
+      # https://github.com/actions/cache/blob/main/examples.md#node---yarn
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+
+      - name: Cache yarn
+        uses: actions/cache@937d24475381cd9c75ae6db12cb4e79714b926ed
+        id: yarn-cache
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: Install dependencies
+        run: yarn install --pure-lockfile
+
+      - name: Verify that canonical-data.json files adhere to their json schema
+        run: yarn format-json
+
+      - name: 'Commit formatted code'
+        run: |
+          # Check if there is nothing to commit (i.e. no formatting changes made)
+          if [ -z "$(git status --porcelain)" ]; then
+            echo "Code is already formatted correctly"
+            exit 0
+          fi
+
+          # Setup the git user (required to commit anything)
+          git config --global user.email "github-actions[bot]@users.noreply.github.com"
+          git config --global user.name "github-actions[bot]"
+
+          # Commit the changes made by prettier
+          git add .
+          git commit -m "[CI] Format code"
+          git push
+
+      - name: 'Post acknowledgement that it has formatted the code'
+        continue-on-error: true # Never fail the build if this fails
+        uses: actions/github-script@e3cbab99d3a9b271e1b79fc96d103a4a5534998c
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          script: |
+            github.issues.createComment({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: 'The "Format code" action has finished running.'
+            })
+
+      - name: 'Post reminder to trigger build manually'
```

Comment:
A question for my own understanding: in what use case does this step run?

It seems like when it is a fork, this is catered for (e.g. ln # 55 - 58), and there is nothing stopping the other steps running when it is a fork... yet this step seems to imply that the job should not work for a fork?

Reply:
It actually says it in the message!

> For security reasons, `/format` does not trigger CI builds when the PR has been submitted from a fork. If checks were not passing due to code format, trigger a build to make the required checks pass, through one of the following ways:
> 
> - Push an empty commit to this branch: `git commit --allow-empty -m "Trigger builds"`.
> - Close and reopen the PR.
> - Push a regular commit to this branch.

Basically when you _write_ commits (so you use the SECRET token, which we do in the format _write_ action) the CI doesn't trigger. So it's not that you cannot run this action on a forked-pr (you can!), it's that _after_ running this action, it won't re-trigger a CI run from the commit that this action writes.

Reply:
> Basically when you write commits (so you use the SECRET token, which we do in the format write action) the CI doesn't trigger. So it's not that you cannot run this action on a forked-pr (you can!), it's that after running this action, it won't re-trigger a CI run from the commit that this action writes.

This is only true for commits written using the default GitHub Actions token. When the PR is from a branch within this repo, we can use an SSH deploy key to push to it and avoid this issue. That's why the workflow has two cases that are handled differently (fork vs non-fork).

Reply:
Oh yeah, there's that :)

Reply:
Great, thanks for clarifying folks!